subt/ros/proxy/traversability.cc

```diff
@@ -0,0 +1,773 @@
+  {
+    bool obscured_view = false;
+    bool ground_visible = false;
+
+    float expected_ground = robot_xyz.z();
+    cv::Mat full_size_uv(3, 1, CV_32F);
+    // As if there was an obstacle visible at the bottom line of the image,
+    // config_.synthetic_obstacle_distance far.
+    full_size_uv.at<float>(0, 0) = u * config_.depth_image_stride;
+    full_size_uv.at<float>(1, 0) = undistorted_img.rows * config_.depth_image_stride - 1; 
+    full_size_uv.at<float>(2, 0) = 1;
+    const auto pt_down_near = *sensor_pose * (config_.synthetic_obstacle_distance * MatToVector3(camera_p_inv * full_size_uv));
+
+    for (int v = undistorted_img.rows - vertical_pixel_offset - 1; v >= 0; --v)
+    {
+      if (!horizontal.at<uint8_t>(v, u) &&
+          !pruned_vertical.at<uint8_t>(v, u))
+      {
+        continue;
+      }
+
+      // Ignore points far above where we expect ground.
+      const cv::Point3f pt = xyz.at<cv::Point3f>(v, u);
+      if (pt.z > expected_ground + config_.robot_height)
+      {
+        continue;
+      }
+
+      const float pt_dist_from_sensor =
+        std::hypot(pt.x - sensor_xyz.x(), pt.y - sensor_xyz.y());
+
+      // Vertical obstacle.
+      if (pruned_vertical.at<uint8_t>(v, u))
+      {
+        // Vertical surfaces in narrow dips in the surface can be ignored,
+        // because the robot can cross the dip. A rail sunk under the surface
+        // is an example of such surface.
+        if (pt.z - expected_ground > config_.max_dip_up ||
+            pt.z - expected_ground < config_.max_dip_down)
+        {
+          if (pt_dist_from_sensor <= config_.visible_ground_min)
+          {
+            obscured_view = true;
+          }
+          if (cnt++ % (config_.depth_subsampling / config_.depth_image_stride / config_.depth_image_stride) == 0)
+          {
+            points.emplace_back(pt.x, pt.y, pt.z);
+          }
+        }
+      }
+      else if (horizontal.at<uint8_t>(v, u))
+      {
+        if (pt.z - robot_xyz.z() > config_.max_bump_height &&
+            pt_dist_from_sensor <= config_.visible_ground_min)
+        {
+          obscured_view = true;
+        }
+
+        if (pt_dist_from_sensor >= config_.visible_ground_min &&
+            pt_dist_from_sensor <= config_.visible_ground_max)
+        {
+          ground_visible = true;
+        }
+
+        expected_ground = pt.z;
+      }
+    }
+
+    // We react to missing ground only when the robot is looking sufficiently
+    // horizontally or downwards. Otherwise, the robot may be climbing up a
+    // ridge and does not even have a chance to see the ground.
+    if (pt_down_near.z() - robot_xyz.z() <= 0) {
+      // If
+      // a) We haven't seen any ground in this direction;
+      // b) The view is not obstructed by an obstacle;
+      // c) The robot is not pitched/rolled in a way that seeing
+      //    ground in this direction is impossible;
+      // then we should be worried.
+      if (!ground_visible)
```

Comment:
This whole debugging block should be deleted.

Reply:
fixed.